source/aws_imds_client.c

```diff
@@ -607,17 +626,61 @@ static void s_query_complete(struct imds_user_data *user_data) {
         return;
     }
 
-    /* In this case we fallback to the secure imds flow. */
     if (user_data->status_code == AWS_HTTP_STATUS_CODE_401_UNAUTHORIZED) {
-        s_invalidate_cached_token_safely(user_data);
-        s_reset_scratch_user_data(user_data);
-        aws_retry_token_release(user_data->retry_token);
-        if (s_get_resource_async_with_imds_token(user_data)) {
-            s_user_data_release(user_data);
+        struct aws_imds_client *client = user_data->client;
+        aws_mutex_lock(&client->token_lock);
+        if (aws_byte_buf_eq(&user_data->imds_token, &client->cached_token)) {
+            /* If the token used matches the cached token, that means the cached token is invalid. */
+            client->token_state = AWS_IMDS_TS_INVALID;
+            AWS_LOGF_DEBUG(
+                AWS_LS_IMDS_CLIENT,
+                "(id=%p) IMDS client's cached token is set to be invalid by requester %p.",
+                (void *)client,
+                (void *)user_data);
+        }
+        /* Flip the flag to let following requests use token as it's required. */
+        client->token_required = true;
+        aws_mutex_unlock(&client->token_lock);
+
+        if (!user_data->imds_token_required) {
```

Comment:
code clarity: combine the error cases into one with a comment
```suggestion
        if (!user_data->imds_token_required && !user_data->is_fallback) {
        	// falback to v1
        } else {
        	// error out 
        
        }
```